patches/flysystem_s3_8.x-1.0-beta1_enable_pre-signed_downloads.patch

```diff
@@ -0,0 +1,426 @@
+   }
+
+@@ -110,6 +142,8 @@ class S3Test extends UnitTestCase {
+     $container = new ContainerBuilder();
+     $container->set('request_stack', new RequestStack());
+     $container->get('request_stack')->push(Request::create('http://example.com/'));
++    $container->set('renderer', $this->renderer);
++    $container->set('logger.channel.flysystem', $this->logger);
+
+     $configuration = [
+       'key'             => 'foo',
+@@ -121,7 +155,6 @@ class S3Test extends UnitTestCase {
+     ];
+
+     $plugin = S3::create($container, $configuration, '', '');
+-    $this->assertSame('http://storage.example.com/my-bucket/foo%201.html', $plugin->getExternalUrl('s3://foo 1.html'));
+     $this->assertSame('https://api.somewhere.tld', (string) $plugin->getAdapter()->getClient()->getEndpoint());
+   }
+
+@@ -131,19 +164,76 @@ class S3Test extends UnitTestCase {
+       'bucket'   => 'my-bucket',
+     ];
+
+-    $plugin = new S3($this->getMock(S3ClientInterface::class), new Config($configuration));
++    $client = $this->prophesize(S3Client::class);
++    $client->getCommand('getObjectAcl', Argument::type('array'))->willReturn($this->prophesize(\Aws\Command::class)->reveal());
++    $client->execute(Argument::type('\Aws\Command'))->willReturn(new \Aws\Result([
++      'Grants' => [
++        ['Grantee' => ['URI' => AwsS3Adapter::PUBLIC_GRANT_URI], 'Permission' => 'READ'],
++      ],
++    ]));
++    $client = $client->reveal();
++    $plugin = new S3($client, new Config($configuration), $this->renderer, $this->logger);
++    $this->assertSame('http://s3.amazonaws.com/my-bucket/foo.html', $plugin->getExternalUrl('s3://foo.html'));
++  }
++
++  /**
++   * Test presigned URL generation.
++   */
++  public function testPresignedUrl() {
++    $configuration = [
++      'bucket'   => 'my-bucket',
++      'expires' => '+10 seconds',
++    ];
++
++    $client = $this->prophesize(S3Client::class);
++    $client->getCommand('getObjectAcl', Argument::type('array'))->willReturn($this->prophesize(\Aws\Command::class)->reveal());
++    $client->execute(Argument::type('\Aws\Command'))->willReturn(new \Aws\Result([
++      'Grants' => [
++        [],
++      ],
++    ]));
++    $client->getCommand('getObject', Argument::type('array'))->willReturn($this->prophesize(\Aws\Command::class)->reveal());
++    $request = new Psr7\Request('GET', 'https://s3.amazonaws.com/signed');
++    $client->createPresignedRequest(Argument::type('\Aws\Command'), '+10 seconds')->willReturn($request);
++    $client = $client->reveal();
++    $kill_switch = new KillSwitch();
++    $plugin = new S3($client, new Config($configuration), $this->renderer, $this->logger, $kill_switch);
++    $this->assertSame('https://s3.amazonaws.com/signed', $plugin->getExternalUrl('s3://foo.html'));
++    $this->assertEquals(ResponsePolicyInterface::DENY, $kill_switch->check(new Response(), new Request()));
++  }
```

Comment:
This patch introduces a bunch of tests 🎉. How can we run these tests (if we don't already)? Could we run this as part of the Docker build or CI?

Reply:
Will spin up a card for this